app/src/main/java/com/hedvig/app/feature/genericauth/GenericAuthViewModel.kt

```diff
@@ -0,0 +1,91 @@
+package com.hedvig.app.feature.genericauth
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.hedvig.app.feature.embark.EMAIL_REGEX
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class GenericAuthViewModel(
+    private val createOtpAttemptUseCase: CreateOtpAttemptUseCase,
+) : ViewModel() {
+    private val _viewState = MutableStateFlow(
+        ViewState(
+            input = "",
+            error = ViewState.TextFieldError.EMPTY,
+            dirty = false,
+            touched = false,
+        )
+    )
+    val viewState = _viewState.asStateFlow()
+
+    data class ViewState(
+        val input: String,
+        val error: TextFieldError?,
+        val dirty: Boolean,
+        val touched: Boolean,
+    ) {
+        enum class TextFieldError {
+            EMPTY,
+            INVALID_EMAIL,
+        }
+    }
+
+    fun setInput(value: String) {
+        _viewState.update { previousState ->
+            previousState.copy(
+                input = value,
+                dirty = true,
+                error = computeError(previousState, value),
+            )
+        }
+    }
+
+    fun clear() {
+        _viewState.update { it.copy(input = "", error = null, dirty = false, touched = false) }
+    }
+
+    fun blur() {
+        _viewState.update { previousState ->
+            val newState = previousState.copy(touched = previousState.dirty)
+            val error = computeError(newState, newState.input)
+            newState.copy(error = error)
+        }
+    }
+
+    fun submitEmail() {
+        viewModelScope.launch {
+            // TODO: Set Loading-state, once one exists
+            val email = viewState.value.input
+            when (val result = createOtpAttemptUseCase.invoke(email)) {
+                is CreateOtpAttemptUseCase.Result.Success -> {
+                    // TODO: Show the OTP input screen, maintain the token in the ViewModel
+                }
+                CreateOtpAttemptUseCase.Result.Error -> {
+                    _viewState.update { it.copy(error = ViewState.TextFieldError.INVALID_EMAIL) }
```

Comment:
This should not be invalid email I suppose?

Reply:
True - perhaps we could request some kind of `network error`-copy from Hiro.

Reply:
Come to think of it, don't we have a generic network error-translation we could utilize here? @hugokallstrom